### PR TITLE
Support reflogs

### PR DESCRIPTION
**Pull request by philip-peterson**
This PR adds support for reflogs on all repositories. It does this by adding a global configuration entry.

Implements #14865

### CODE REVIEW

File: modules/git/repo.go
Comment (philip-peterson) [inline modules/git/repo.go]: I did not fully understand the purpose of deleting instead of just writing, but left it here since that's what the hooks task does as well.

File: services/cron/tasks_extended.go
Comment (philip-peterson) [inline services/cron/tasks_extended.go]: Seems as though maybe this should run at start, however the same can be said of hooks, so I left it be.

File: modules/git/repo.go
Comment (zeripath) [reply modules/git/repo.go]: Argh is the hook task writing directly to .gitconfig too?

You must not do this. It's fine to do that if you're managing .gitconfig by hand on a few repos but there will be concurrency issues if you're doing it on a server.

Use git config to write to .gitconfig instead.

**Comment by zeripath** [comment]
We must not write directly to .gitconfig on repositories. Use the git config command instead as it will lock things properly.

File: modules/git/repo.go
Comment (philip-peterson) [reply modules/git/repo.go]: Hooks task simply writes hook scripts. I am not sure what issues specifically that locking would prevent, but I have a feeling that whatever it is, the hooks task would have the same issue.

At the very least, it seems like deleting hooks before re-writing them would cause concurrency issues if the hook was accessed during the moment in which it's deleted, I think you could at least eliminate that by just opening the file for writing instead of deleting first.

**Comment by philip-peterson** [comment]
Thanks for the review, I will add in some explicit git config commands there.

File: modules/git/repo.go
Comment (zeripath) [reply modules/git/repo.go]: ok so the hook tasks aren't writing to .git/config ... that's fine(ish) (There is a possible concurrency issue there but it's actually less of a problem.)

However, in this case you must use git config to adjust the config.

Comment (philip-peterson) [reply modules/git/repo.go]: I have no problem with that. I do have a question though, so assuming I only run a git config command, I probably will not want to fully sync the `gitconfigs` anymore and simply just change these two options (the reflog settings). Or should I try to be aware of all possible configuration settings since there may be some future use of these?

Alternatively I could even hoist these up to a more global gitconfig, so we don't always need to be modifying the individual repos.

Comment (zeripath) [reply modules/git/repo.go]: Yup, just setting the .gitconfig would work and we could that without a even a cron task - just set it on init.

Comment (zeripath) [reply modules/git/repo.go]: BTW I've just re-read what you're doing here and this code would have resulted in destroying mirrors!!

Comment (philip-peterson) [reply modules/git/repo.go]: Yep, sorry about that. First PR and all... I'm not sure what all the capabilities are yet so it will be a bit of a journey to avoid stepping on toes.

**Review by techknowlogick — DISMISSED** [review]
Thanks for this PR!

Setting this as blocked due to @zeripath's comment. Mergers, feel free to dismiss my review.

File: modules/git/repo.go
Comment (philip-peterson) [reply modules/git/repo.go]: OK, gitconfigs are no longer written, it's just stored in the global gitconfig now.

File: modules/git/git.go
Comment (philip-peterson) [inline modules/git/git.go:244]: Is there an easy way to check what versions of Git support this new feature? A testing matrix or anything? Reflogs are pretty old so I was having trouble figuring out what the oldest version of Git that supports them is.

Comment (delvh) [reply modules/git/git.go:244]: I just had a look at the git repo with a `git blame`, I've been able to trace reflogs back to Git 1.5.
As the minimum supported Git version is 2.0 there is nothing to worry about in this case.

File: custom/conf/app.example.ini
Comment (jtran) [inline custom/conf/app.example.ini:2284]: All new config settings should be added to the cheat sheet documentation.  See `docs/content/doc/advanced/config-cheat-sheet.en-us.md`.

Comment (philip-peterson) [reply custom/conf/app.example.ini:2284]: Done!

**Comment by philip-peterson** [comment]
@techknowlogick I think this should be unblocked now.

**Review by kdumontnu — CHANGES_REQUESTED** [review]
I really like this PR, but I think some of the supporting functions appear to be broken. When I set:
```
[git.reflog]
ENABLED = false
```
I get `routers/init.go:70:mustInitCtx() [F] code.gitea.io/gitea/modules/git.InitFull(ctx) failed: failed to unset git global config core.logAllRefUpdates, err: exit status 5` and the startup fails.

I think this is due to an issue with [`configUnsetAll`](https://github.com/go-gitea/gitea/blob/8540fc45b11eff9a73753ca139f8ea5c38509bf5/modules/git/git.go#L368), where it first "get"s the value without any config flags (so it will check all config settings), but then it "unset"s using the `--global` flag and panics when that key isn't found.

The logic was updated in https://github.com/go-gitea/gitea/pull/19732. It seems to me that we should be consistent (note: the configSet function follows similar logic). @wxiaoguang what was the intention with that logic? I imagine we always want to be getting/setting the user config (`--global`) settings with those functions. 

I appear to have a similar issue when I set `ENABLED = true`, but in that case it doesn't error, I just can't seem to get the value to work. If I run `git config --get core.logAllRefUpdates` from one of my repo directories I get exit(1).

It's possible this is due to my local settings - I think somehow git is seeing my local `gitea` repo config settings when it runs `configSet`/`configUnsetAll`.

**Comment by wxiaoguang** [comment]
> The logic was updated in #19732. It seems to me that we should be consistent (note: the configSet function follows similar logic). @wxiaoguang what was the intention with that logic? I imagine we always want to be getting/setting the user config (`--global`) settings with those functions.

That logic is as old as #14295 , https://github.com/go-gitea/gitea/pull/14295/files#diff-e4a61ced421340955369c845734686fe5acd82c4713915f8f3d26b7c95cf9f4eR273-R294

I'm not sure whether it is right. If I remember correctly, I didn't touch the logic too much when doing refactoring.

**Comment by philip-peterson** [comment]
We can just remove that check then. What could possibly go wrong?

**Comment by philip-peterson** [comment]
Jokes aside, I think the issue is that the config-get is not global, while the config-unset-all is. From what I understand, if you try to config-unset-all something that's not set, you'll get an exit code 5, i.e. not success. Hence the check. I guess we can modify that to be --global as well, maybe I'll open another PR for that.

**Review by kdumontnu — APPROVED** [review]
> @kdumontnu are you able to confirm that code change would fix the issue?

> Edit: my bad, I didn't realize I just said exactly what @kdumontnu said. I concur though with @kdumontnu.

No worries, just happy we came to the same conclusion. Thanks for the patch!

I rebased on #23157 and it works great. Just to note that to test locally I have to run `HOME=/home/kyle/1-develop/gitea/data/home/ git reflog show --all` since PR #19732

Probably want to wait to merge this until #23157 is merged

**Comment by philip-peterson** [comment]
OK, I am not sure I follow the HOME thing, but let me know if there are any more PRs I can open. Happy to help get this and the other one through.

**Comment by kdumontnu** [comment]
Adding 1.19 backport (might want 1.18 too)

File: modules/git/git.go
Comment (lunny) [inline modules/git/git.go]: Why not just Update it? Is the check necessary?

Comment (philip-peterson) [reply modules/git/git.go]: I figured it was nice to keep the .gitconfig as minimal as possible. It's not necessary, no.

Comment (wxiaoguang) [reply modules/git/git.go]: I guess it would cause bug if `setting.Git.Reflog.Expiration==90`, but `gc.reflogExpire != 90` in config

Comment (lunny) [reply modules/git/git.go]: I think this PR assumes the default value of git reflog expire is always 90?

Comment (wxiaoguang) [reply modules/git/git.go]: What if a user:

* Set Expiration=14 , write reflogExpire=14 to config
* Change Expiration to 90
* Then the reflogExpire in config is still 14?

Comment (philip-peterson) [reply modules/git/git.go]: The config is synced whenever `GlobalInitInstalled` is invoked, via `git.InitFull`. From what I understand, this should happen whenever the server restarts. This setting is therefore like others in `app.ini`, in that they are static for the lifetime of the Gitea process and will be refreshed when the process is restarted.

Comment (philip-peterson) [reply modules/git/git.go]: Removed the 90 days check.

**Review by lunny — APPROVED** [review]

File: modules/git/git.go
Comment (wxiaoguang) [reply modules/git/git.go]: Yup, I guess the old code might work like this:

* Set Expiration=14 in `app.ini`, start Gitea, write reflogExpire=14 to git config
* Stop Gitea, Change Expiration in `app.ini` to 90
* Start Gitea, `if Expiration != 90` check fails, then the reflogExpire in git config is still 14, it's inconsistent with the value in `app.ini`

I think removing the check is right (as the new code)

Comment (philip-peterson) [reply modules/git/git.go]: Ah I see, so the issue was I wasn't unsetting. Noted, yeah that's what I get for trying to make things simpler, I made it more complicated instead.

**Comment by philip-peterson** [comment]
PR should be good to go.

File: config/app.ini
Comment (lunny) [inline config/app.ini]: It's pushed accidently?

Comment (philip-peterson) [reply config/app.ini]: great catch, thank you. fixed.

**Comment by lunny** [comment]
@techknowlogick needs your approval.

**Review by zeripath — APPROVED** [review]

**Pull request merged**
